### PR TITLE
Fix e2e test failure for RecreateDeployment

### LINKED ISSUE #19732: e2e flake: RecreateDeployment should delete old pods and create new ones [reaper]
See https://github.com/kubernetes/kubernetes/pull/19638#issuecomment-171832286

> Error logs:
```
Jan 14 16:50:43.853: INFO: event for nginx-controller-f5u6e: {default-scheduler } Scheduled: Successfully assigned nginx-controller-f5u6e to 127.0.0.1
Jan 14 16:50:43.853: INFO: event for nginx-controller-f5u6e: {kubelet 127.0.0.1} Pulling: pulling image "nginx"
Jan 14 16:50:43.853: INFO: event for nginx-controller-f5u6e: {kubelet 127.0.0.1} Pulled: Successfully pulled image "nginx"
Jan 14 16:50:43.853: INFO: event for nginx-controller-f5u6e: {kubelet 127.0.0.1} Created: Created container with docker id 80c7021dd353
Jan 14 16:50:43.853: INFO: event for nginx-controller-f5u6e: {kubelet 127.0.0.1} Started: Started container with docker id 80c7021dd353
Jan 14 16:50:43.853: INFO: event for nginx-controller-f5u6e: {kubelet 127.0.0.1} Killing: Killing container with docker id 80c7021dd353: Need to kill pod.
Jan 14 16:50:43.853: INFO: event for nginx-controller-n4v05: {default-scheduler } Scheduled: Successfully assigned nginx-controller-n4v05 to 127.0.0.1
Jan 14 16:50:43.853: INFO: event for nginx-controller-n4v05: {kubelet 127.0.0.1} Pulling: pulling image "nginx"
Jan 14 16:50:43.853: INFO: event for nginx-controller-n4v05: {kubelet 127.0.0.1} Pulled: Successfully pulled image "nginx"
Jan 14 16:50:43.853: INFO: event for nginx-controller-n4v05: {kubelet 127.0.0.1} Created: Created container with docker id c7a6f514bbce
Jan 14 16:50:43.853: INFO: event for nginx-controller-n4v05: {kubelet 127.0.0.1} Started: Started container with docker id c7a6f514bbce
Jan 14 16:50:43.853: INFO: event for nginx-controller-n4v05: {kubelet 127.0.0.1} Killing: Killing container with docker id c7a6f514bbce: Need to kill pod.
Jan 14 16:50:43.853: INFO: event for nginx-controller-z5vvs: {default-scheduler } Scheduled: Successfully assigned nginx-controller-z5vvs to 127.0.0.1
Jan 14 16:50:43.853: INFO: event for nginx-controller-z5vvs: {kubelet 127.0.0.1} Pulling: pulling image "nginx"
Jan 14 16:50:43.853: INFO: event for nginx-controller-z5vvs: {kubelet 127.0.0.1} Pulled: Successfully pulled image "nginx"
Jan 14 16:50:43.853: INFO: event for nginx-controller-z5vvs: {kubelet 127.0.0.1} Created: Created container with docker id 8458aaa829b1
Jan 14 16:50:43.853: INFO: event for nginx-controller-z5vvs: {kubelet 127.0.0.1} Started: Started container with docker id 8458aaa829b1
Jan 14 16:50:43.853: INFO: event for nginx-controller-z5vvs: {kubelet 127.0.0.1} Killing: Killing container with docker id 8458aaa829b1: Need to kill pod.
Jan 14 16:50:43.853: INFO: event for nginx-controller: {replication-controller } SuccessfulCreate: Created pod: nginx-controller-f5u6e
Jan 14 16:50:43.853: INFO: event for nginx-controller: {replication-controller } SuccessfulCreate: Created pod: nginx-controller-n4v05
Jan 14 16:50:43.853: INFO: event for nginx-controller: {replication-controller } SuccessfulCreate: Created pod: nginx-controller-z5vvs
Jan 14 16:50:43.853: INFO: event for nginx-controller: {replication-controller } SuccessfulDelete: Deleted pod: nginx-controller-n4v05
Jan 14 16:50:43.853: INFO: event for nginx-controller: {replication-controller } SuccessfulDelete: Deleted pod: nginx-controller-z5vvs
Jan 14 16:50:43.853: INFO: event for nginx-controller: {replication-controller } SuccessfulDelete: Deleted pod: nginx-controller-f5u6e
Jan 14 16:50:43.853: INFO: event for redis-deployment-3-io8ah-gwz5j: {default-scheduler } Scheduled: Successfully assigned redis-deployment-3-io8ah-gwz5j to 127.0.0.1
Jan 14 16:50:43.853: INFO: event for redis-deployment-3-io8ah-gwz5j: {kubelet 127.0.0.1} Pulling: pulling image "redis"
Jan 14 16:50:43.853: INFO: event for redis-deployment-3-io8ah-gwz5j: {kubelet 127.0.0.1} Pulled: Successfully pulled image "redis"
Jan 14 16:50:43.853: INFO: event for redis-deployment-3-io8ah-gwz5j: {kubelet 127.0.0.1} Created: Created container with docker id 169f1a863593
Jan 14 16:50:43.853: INFO: event for redis-deployment-3-io8ah-gwz5j: {kubelet 127.0.0.1} Started: Started container with docker id 169f1a863593
Jan 14 16:50:43.853: INFO: event for redis-deployment-3-io8ah-mzt95: {default-scheduler } Scheduled: Successfully assigned redis-deployment-3-io8ah-mzt95 to 127.0.0.1
Jan 14 16:50:43.853: INFO: event for redis-deployment-3-io8ah-mzt95: {kubelet 127.0.0.1} Pulling: pulling image "redis"
Jan 14 16:50:43.853: INFO: event for redis-deployment-3-io8ah-mzt95: {kubelet 127.0.0.1} Pulled: Successfully pulled image "redis"
Jan 14 16:50:43.853: INFO: event for redis-deployment-3-io8ah-mzt95: {kubelet 127.0.0.1} Created: Created container with docker id 3173680b59fa
Jan 14 16:50:43.853: INFO: event for redis-deployment-3-io8ah-mzt95: {kubelet 127.0.0.1} Started: Started container with docker id 3173680b59fa
Jan 14 16:50:43.853: INFO: event for redis-deployment-3-io8ah-wfoti: {default-scheduler } Scheduled: Successfully assigned redis-deployment-3-io8ah-wfoti to 127.0.0.1
Jan 14 16:50:43.853: INFO: event for redis-deployment-3-io8ah-wfoti: {kubelet 127.0.0.1} Pulling: pulling image "redis"
Jan 14 16:50:43.853: INFO: event for redis-deployment-3-io8ah-wfoti: {kubelet 127.0.0.1} Pulled: Successfully pulled image "redis"
Jan 14 16:50:43.853: INFO: event for redis-deployment-3-io8ah-wfoti: {kubelet 127.0.0.1} Created: Created container with docker id 808f7a9b408a
Jan 14 16:50:43.853: INFO: event for redis-deployment-3-io8ah-wfoti: {kubelet 127.0.0.1} Started: Started container with docker id 808f7a9b408a
Jan 14 16:50:43.854: INFO: event for redis-deployment-3-io8ah: {replication-controller } SuccessfulCreate: Created pod: redis-deployment-3-io8ah-gwz5j
Jan 14 16:50:43.854: INFO: event for redis-deployment-3-io8ah: {replication-controller } SuccessfulCreate: Created pod: redis-deployment-3-io8ah-wfoti
Jan 14 16:50:43.854: INFO: event for redis-deployment-3-io8ah: {replication-controller } SuccessfulCreate: Created pod: redis-deployment-3-io8ah-mzt95
Jan 14 16:50:43.854: INFO: event for redis-deployment-3: {deployment-controller } ScalingRC: Scaled down rc nginx-controller to 0
Jan 14 16:50:43.854: INFO: event for redis-deployment-3: {deployment-controller } ScalingRC: Scaled up rc redis-deployment-3-io8ah to 3
Jan 14 16:50:44.047: INFO: POD                             NODE       PHASE    GRACE  CONDITIONS
Jan 14 16:50:44.047: INFO: nginx-deployment-w97x3-b1vyz    127.0.0.1  Running         [{Ready True 0001-01-01 00:00:00 +0000 UTC 2016-01-14 16:42:09 -0800 PST  }]
Jan 14 16:50:44.047: INFO: nginx-deployment-w97x3-t1fw3    127.0.0.1  Running         [{Ready True 0001-01-01 00:00:00 +0000 UTC 2016-01-14 16:42:05 -0800 PST  }]
Jan 14 16:50:44.047: INFO: nginx-deployment-w97x3-u5a6k    127.0.0.1  Running         [{Ready True 0001-01-01 00:00:00 +0000 UTC 2016-01-14 16:42:07 -0800 PST  }]
Jan 14 16:50:44.047: INFO: redis-deployment-3-io8ah-gwz5j  127.0.0.1  Running         [{Ready True 0001-01-01 00:00:00 +0000 UTC 2016-01-14 16:45:47 -0800 PST  }]
Jan 14 16:50:44.047: INFO: redis-deployment-3-io8ah-mzt95  127.0.0.1  Running         [{Ready True 0001-01-01 00:00:00 +0000 UTC 2016-01-14 16:45:50 -0800 PST  }]
Jan 14 16:50:44.047: INFO: redis-deployment-3-io8ah-wfoti  127.0.0.1  Running         [{Ready True 0001-01-01 00:00:00 +0000 UTC 2016-01-14 16:45:48 -0800 PST  }]
```
```  
  Expected error:
      <*errors.errorString | 0xc8201a1d10>: {
          s: "timed out waiting for the condition",
      }
      timed out waiting for the condition
  not to have occurred
```

### PR DESCRIPTION
**Pull request by janetkuo**
Fixes #19732. 

When a deployment has `RecreateDeployment` strategy, it doesn't have `.Spec.Strategy.RollingUpdate.MinReadySeconds`, which makes calculating available pods panic. 

Fix it by making pass MinReadySeconds=0 at RecreateDeployment case. 

cc @mqliang @nikhiljindal @ghodss @ironcladlou @kargakis @bgrant0607

### CODE REVIEW

File: pkg/controller/deployment/deployment_controller.go [hunk 1]
```diff
@@ -709,24 +707,43 @@ func (dc *DeploymentController) scaleUpNewRCForRecreate(newRC *api.ReplicationCo
...
+	updatedReplicas = deploymentutil.GetReplicaCountForRCs([]*api.ReplicationController{newRC})
+	switch deployment.Spec.Strategy.Type {
+	case extensions.RecreateDeploymentStrategyType:
+		availableReplicas, err = deploymentutil.GetAvailablePodsForRCs(dc.client, allRCs, 0)
+		if err != nil {
+			err = fmt.Errorf("failed to count available pods: %v", err)
+			return
+		}
+	case extensions.RollingUpdateDeploymentStrategyType:
+		availableReplicas, err = deploymentutil.GetAvailablePodsForRCs(dc.client, allRCs, deployment.Spec.Strategy.RollingUpdate.MinReadySeconds)
```
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: If the two cases are identical except for the last parameter, I suggest just setting a variable in the switch and moving the call to GetAvailablePodsForRCs below.

Comment (janetkuo) [inline pkg/controller/deployment/deployment_controller.go]: Done

**Comment by janetkuo** [comment]
@bgrant0607 Comments addressed. PTAL.

**Comment by bgrant0607** [comment]
LGTM, thanks

@k8s-oncall This is to fix an e2e test failure.

**Comment by kargakis** [comment]
Fundamentaly, this is still wrong code, right? When the controller decides at the very start of each deployment reconcilation the path that will be followed based on the deployment strategy, then it shouldn't need to make such a decision later on. I believe the correct change here would be to separate the paths and use strategy-specific `syncDeploymentStatus` units.

**Comment by janetkuo** [comment]
@k8s-bot e2e test this

**Comment by janetkuo** [comment]
@kargakis I'm not sure if I'm not following you. `syncDeploymentStatus` calls `calculateStatus` which returns result (deployment status) based on different strategy. The reason that we call `syncDeploymentStatus` is because the update of deployment status after scaling up/down may sometime fail (for example, when the deployment is changed when we try to update its status), and we need to periodically sync its status.

**Comment by bgrant0607** [comment]
API-wise, I'd be in favor of moving MinReadySeconds from RollingUpdateDeployment to DeploymentStrategy or DeploymentSpec, since it could be used to determine which pods are "available" regardless of the strategy.

**Comment by bgrant0607** [comment]
But that shouldn't block this PR, since it fixes an e2e test failure.

**Comment by bgrant0607** [comment]
Manually merging.

**Pull request merged**